### PR TITLE
Groovy test listener

### PR DESCRIPTION
**Pull request by MykolaGolubyev**
@tsiq-clemens this is so you can do post test run validation for your metadata values. Or automatically inject metadata values based on test names.

### CODE REVIEW

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListeners.java [hunk 1]
```diff
@@ -2,43 +2,68 @@
...
 import com.twosigma.webtau.utils.ServiceLoaderUtils;
 
+import java.util.ArrayList;
 import java.util.List;
+import java.util.function.Supplier;
+import java.util.stream.Stream;
 
 public class TestListeners {
-    private static List<TestListener> listeners = ServiceLoaderUtils.load(TestListener.class);
+    private static List<TestListener> discoveredListeners = ServiceLoaderUtils.load(TestListener.class);
```
Comment (tsiq-karold) [inline webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListeners.java]: Should these three `private static` variables be a) final and b) in caps?

File: webtau-docs/webtau/groovy-specific-runner/test-listeners.md [hunk 2]
```diff
@@ -0,0 +1,13 @@
+# Test Listener Interface
+
+To define a custom test listener you need to implement `TestListener` interface
```
Comment (tsiq-karold) [inline webtau-docs/webtau/groovy-specific-runner/test-listeners.md]:
```suggestion
To define a custom test listener you need to implement the `TestListener` interface
```

File: webtau-docs/webtau/groovy-specific-runner/test-listeners.md [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+:include-file: com/twosigma/webtau/reporter/TestListener.java {title: "TestListener.java"}
+
+# Config
+
+Once you have an implementation, use `testListeners` key in config file to use it
```
Comment (tsiq-karold) [inline webtau-docs/webtau/groovy-specific-runner/test-listeners.md]:
```suggestion
Once you have an implementation, use the `testListeners` key in your config file to use it
```

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListeners.java (hunk 1 above)
Comment (MykolaGolubyev) [reply webtau-core/src/main/java/com/twosigma/webtau/reporter/TestListeners.java]: final yes, changing. 
don't want CAPS, as they are not actually constants per se.

File: webtau-docs/webtau/groovy-specific-runner/test-listeners.md (hunk 3 above)
Comment (MykolaGolubyev) [reply webtau-docs/webtau/groovy-specific-runner/test-listeners.md]: thanks

File: webtau-core/src/main/java/com/twosigma/webtau/reporter/WebTauTestMetadata.java [hunk 4]
```diff
@@ -45,6 +45,20 @@ public WebTauTestMetadata() {
         return previousMetaValues;
     }
 
+    public boolean has(String key) {
+        return meta.containsKey(key);
```
Comment (tsiq-clemens) [inline webtau-core/src/main/java/com/twosigma/webtau/reporter/WebTauTestMetadata.java]: s/meta/metadata

File: webtau-docs/webtau/groovy-specific-runner/test-listeners.md [hunk 5]
```diff
@@ -0,0 +1,13 @@
+# Test Listener Interface
```
Comment (tsiq-clemens) [inline webtau-docs/webtau/groovy-specific-runner/test-listeners.md]: this is very helpful, thanks

File: webtau-groovy/src/main/groovy/com/twosigma/webtau/cfg/WebTauGroovyFileConfigHandler.groovy [hunk 6]
```diff
@@ -100,13 +103,22 @@ class WebTauGroovyFileConfigHandler implements WebTauConfigHandler {
...
+
+    private static <E> List<E> instancesFromConfig(ConfigObject config, String key) {
+        def classes = (List<Class<E>>) config.get(key)
+        if (!classes) {
+            return []
         }
 
-        def handlerInstances = handlerClasses.collect { handlerClass -> constructFromClass(handlerClass) }
-        handlerInstances.each { PageElementGetSetValueHandlers.add((PageElementGetSetValueHandler) it) }
+        return classes.collect{ c -> (E)constructFromClass(c) }
```
Comment (tsiq-clemens) [inline webtau-groovy/src/main/groovy/com/twosigma/webtau/cfg/WebTauGroovyFileConfigHandler.groovy]: minor nit: space after `(E)`

**Comment by tsiq-clemens** [comment]
Looks good to me overall, I especially appreciate the code examples.

**Pull request merged**